keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1158 @@
++              set: default
++              path: ca_certificates.pem
+```
+
+When my client pod is scheduled onto a node, Kubelet writes the chosen trust
+anchor set into the container filesystem of my main container.  My main
+container doesn't start until `ca_certificates.pem` is present and non-empty.
+
+I ensure that my client application is capable of reading a set of trust anchors
+from a file containing PEM-wrapped DER-formatted X.509 certificates at the
+location I've specified
+(`/var/run/example-com-server-tls-trust-anchors/ca_certificates.pem`).
+
+I ensure that my client application notices changes to `ca_certificates.pem`
+(using inotify or periodic polling), and properly updates its TLS stack when
+that happens.
+
+#### TrustAnchorSets for Webhook Backends
+
+(To be added)
+
+### Risks and Mitigations
+
+Scalability: In the limit, TrustAnchorSets objects will be used by every pod in
+the cluster, which will require watches from all Kubelets.  When they are
+updated, workloads will need to receive the updates fairly quickly (within 5
+minutes across the whole cluster), to accommodate emergency rotation of trust
+anchors for a private CA.
+
+Security: Should Kubelet attempt to enforce which workloads are allowed to mount
+particular TrustAnchorSets using a trustAnchorSets volume?  The straightforward
+design would have all TrustAnchorSet reads happen using a `system:node:xyz`
+identity, which would allow any workload to mount any TrustAnchorSets object.
+
+Security: Should individual trust anchor set entries designate an OSCP endpoint
+to check for certificate revociation?  Or should we require the URL to be
+embedded in the issued certificates?
+
+Reliability: How can we safely canary a change to a TrustAnchorSets object?
+
+## Design Details
+
+### TrustAnchorSets Object
+
+The TrustAnchorSets object allows signers to publish their trust anchors for
+consumption by workloads, whether directly, via Kubelet, or otherwise.
+
+Every signer corresponds to one TrustAnchorSets object.  TrustAnchorSets objects
+are associated to signers by a naming convention.  It is not required for a
+signer to have a corresponding TrustAnchorSets object, nor is it required for a
+TrustAnchorSets object to correspond to a signer.
+
+A TrustAnchorSets object is associated with a signer if its object name is the
+signer name, after applying the following transformation:
+
+```go
+var trustAnchorSetsName strings.Builder
+for _,c := range signerName {
+  switch {
+  case c == '-':
+    trustAnchorSetsName.WriteString("--")
+  case c == '/':
+    trustAnchorSetsName.WriteString("-slash-")
+  default:
+    trustAnchorSetsName.WriteRune(c)
+  }
+}
+```
+
+For example, the signer name `example.com/my-signer` becomes the TrustAnchorSets
+object name `example.com-slash-my--signer`.
+
+Note that this transformation is one to one and reversible, with `-` being used
+as an escape character.  To convert backwards from a TrustAnchorSets object name
+to a signer name:
```

Comment:
This might be the first time we've minted a conversion like this. If so, might be worth asking SIG Architecture for an opinion on using this mechanism and not an alternative such as the value of a label.

Reply:
If there are two TrustAnchorSets for a signer name, is that a bad thing? For example, this approach:
- the signer name `example.com/my-signer` becomes the TrustAnchorSet object names `green-example.com-slash-my--signer` and `blue-example.com-slash-my--signer`.
- a TrustAnchorSet is immutable once created
- the convention is to start a TrustAnchorSet object name with 1 or more characters that are not `-`, '/' or '.', and that if there is only one TrustAnchorSet for a signer then its prefix should be `default`.
- the prefix is followed by `-` and then the transformed signer name
- signers are welcome to use any other prefixes they like; these could be `blue` and `green`, or a representation of the incept date for the TrustAnchorSet, or anything else.
- signers are encouraged to use predictable, bounded prefixes so that writing RBAC rules becomes feasible. For example, an RBAC rule to cover both the blue and green prefix is trivial to implement.

Reply:
My understanding is that a TrustAnchorSets (or whatever we call it) should never contain intermediates, only roots.  In TLS connection establishment, it's the job of the entity supplying the certificate to its peer to *also* supply any necessary intermiedates to build a chain to one of the peer's roots.

Reply:
OK, fair. Let's document that in the KEP etc so that it's clear that specifying intermediates is a bug / misuse / wahatever.

Reply:
Defining "roots" matters here; there are reasons to put "intermediate" certificates into a trust anchor set, such as CAs that are cross-signed (As was done in the early days of LetsEncrypt). Those CAs were also valid Root certificates in and of themselves.

The definition that matters is that any certificate placed into the trust anchor set  will be treated as a root itself (including, in some badly-behaving TLS implementations, when it doesn't have the CA bit set). Placing your signing intermediate into the set is syntactically valid but would result in your inability to revoke it.

Reply:
OK, I think everything that was discussed in this comment chain is now addressed in the doc.

Canarying: Individual trustAnchorSet volumes now refer to a particular named ClusterTrustBundle.  A given signer can have multiple associated ClusterTrustBundles.  Canarying is accomplished by adding a new ClusterTrustBundle (with a new name) with your desired changes and updating some pods to refer to the new CluterTrustBundle.

Ordering: ClusterTrustBundles now store a single string that contains a PEM bundle.  The field comment makes it clear that entries in the bundle are trust anchors, and they are not considered to be ordered.